CRM/Upgrade/Incremental/Base.php

```diff
@@ -158,6 +158,42 @@ protected function addTask($title, $funcName) {
     $queue->createItem($task, ['weight' => -1]);
   }
 
+  /**
+   * Add a task to activate an extension. This task will run post-upgrade (after all
+   * changes to core DB are settled).
+   *
+   * @param string $title
+   * @param string[] $keys
+   *   List of extensions to enable.
+   */
+  protected function addExtensionTask(string $title, array $keys): void {
+    Civi::queue(CRM_Upgrade_Form::QUEUE_NAME)->createItem(
+      new CRM_Queue_Task([static::CLASS, 'enableExtension'], [$keys], $title),
+      ['weight' => 2000]
+    );
+  }
+
+  /**
+   * @param \CRM_Queue_TaskContext $ctx
+   * @param string[] $keys
+   *   List of extensions to enable.
+   * @return bool
+   */
+  public static function enableExtension(CRM_Queue_TaskContext $ctx, array $keys): bool {
+    // The `enableExtension` has a very high value of `weight`, so this runs after all
+    // core DB schema updates have been resolved. We can use high-level services.
+
+    Civi::dispatcher()->setDispatchPolicy(\CRM_Upgrade_DispatchPolicy::get('upgrade.finish'));
+    $restore = \CRM_Utils_AutoClean::with(function() {
+      Civi::dispatcher()->setDispatchPolicy(\CRM_Upgrade_DispatchPolicy::get('upgrade.main'));
+    });
+
+    $manager = CRM_Extension_System::singleton()->getManager();
+    $manager->enable($manager->findInstallRequirements($keys));
+    CRM_Core_Invoke::rebuildMenuAndCaches(TRUE, TRUE);
```

Comment:
I feel like this line might fix the problems people have reported over the years of the upgrader not completely flushing caches, because this one runs with a full bootstrap. Maybe we should do this unconditionally at the end of every upgrade.

Reply:
We already do! (Per `doCoreFinish()` nee `doFinish()`) This is an additional call just to deal with the extension. (You may have noticed this after posting.)

FWIW, there is a subtle difference, and I'm on the fence about it: 

* `enableExtension()` passes `$triggerRebuild=TRUE`
* `doFinish()` passes `$triggerRebuild=FALSE` and compensates with `fixSchemaDifferences()`

It's also arguably a bit redundant to have two calls. One call (scheduled at, say, `weight=3000` instead of `weight=1000`) might do the job. But I think the pre-conditions/post-conditions are a bit clearer with the two calls - so I'd expect it to be less brittle in the face of future changes.